libtransmission/torrent-queue.h

```diff
@@ -0,0 +1,48 @@
+// This file Copyright © Mnemosyne LLC.
+// It may be used under GPLv2 (SPDX: GPL-2.0-only), GPLv3 (SPDX: GPL-3.0-only),
+// or any future license endorsed by Mnemosyne LLC.
+// License text can be found in the licenses/ folder.
+
+#pragma once
+
+#ifndef __TRANSMISSION__
+#error only libtransmission should #include this header.
+#endif
+
+#include <cstddef>
+#include <string>
+#include <string_view>
+#include <vector>
+
+struct tr_torrent;
+
+class tr_torrent_queue
+{
+public:
+    explicit tr_torrent_queue(std::string_view config_dir)
+        : config_dir_{ config_dir }
+    {
+    }
+    tr_torrent_queue(tr_torrent_queue const&) = delete;
+    tr_torrent_queue(tr_torrent_queue&&) = delete;
+    tr_torrent_queue& operator=(tr_torrent_queue const&) = delete;
+    tr_torrent_queue& operator=(tr_torrent_queue&&) = delete;
+
+    size_t add(tr_torrent const& tor);
+    void remove(tr_torrent const& tor);
+
+    [[nodiscard]] size_t get_pos(tr_torrent const& tor);
+    void set_pos(tr_torrent const& tor, size_t new_pos);
+
+    bool to_file() const;
+    [[nodiscard]] std::vector<std::string> from_file();
+
+    static auto constexpr MinQueuePosition = size_t{};
+    static auto constexpr MaxQueuePosition = ~size_t{};
+
+private:
+    std::vector<tr_torrent const*> queue_;
```

Comment:
It seems like this class doesn't really need to know what a `tr_torrent` is at all. Would it be possible to use `tr_torrent_id_t` instead of `tr_torrent const` as what we track internally?

Having a long-lived `std::vector<tr_torrent const*>` outside of  the `tr_torrents` class seems inherently risky, but that's a secondary issue. Mostly I'm interested in decoupling code that doesn't need to be coupled.

e.g. I would replace every `tr_torrent const&` and `tr_torrent const*` in this header with `tr_torrent_id_t`

The only extra scaffolding we'd need would be a mediator for `store_filename(tr_torrent_id_t)`

Reply:
3cbf8e3fef0293f6272afd8ceb2a84dfac04e63c

Reply:
Thanks!